From 9a67a3e36b01623bf47727c84229ccbacb03b877 Mon Sep 17 00:00:00 2001
From: Tang Guifang <guifang.tang@intel.com>
Date: Fri, 1 Feb 2013 04:14:48 -0500
Subject: [PORT FROM MAIN]OMX: avoid illegal access to NULL pointer in callback.

BZ: 81944

When OMX client gone with exception, in binderDied function it will
remove the corresponding node from mDispatchers array. Callback function
will still go on even if the dispatcher is not found, and causes crash.
Add NULL check to avoid such crash issue.

Change-Id: I43bc25981c8ee9ea30452a9498148be4bd4cc487
Orig-Change-Id: I4ca5542d3cc2195cc4a6f270cafb23a49ba9b8fc
Signed-off-by: Dan Liang <dan.liang@intel.com>
---
 media/libstagefright/omx/OMX.cpp |   21 ++++++++++++++++++---
 1 files changed, 18 insertions(+), 3 deletions(-)

diff --git a/media/libstagefright/omx/OMX.cpp b/media/libstagefright/omx/OMX.cpp
index 3987ead..e231123 100644
--- a/media/libstagefright/omx/OMX.cpp
+++ b/media/libstagefright/omx/OMX.cpp
@@ -414,7 +414,12 @@ OMX_ERRORTYPE OMX::OnEvent(
     msg.u.event_data.data1 = nData1;
     msg.u.event_data.data2 = nData2;
 
-    findDispatcher(node)->post(msg);
+    sp<OMX::CallbackDispatcher> cbd = findDispatcher(node);
+    if (cbd == NULL) {
+        LOGE("CallbackDispatcher for this node is not found");
+        return OMX_ErrorUndefined;
+    }
+    cbd->post(msg);
 
     return OMX_ErrorNone;
 }
@@ -428,7 +433,12 @@ OMX_ERRORTYPE OMX::OnEmptyBufferDone(
     msg.node = node;
     msg.u.buffer_data.buffer = pBuffer;
 
-    findDispatcher(node)->post(msg);
+    sp<OMX::CallbackDispatcher> cbd = findDispatcher(node);
+    if (cbd == NULL) {
+        LOGE("CallbackDispatcher for this node is not found");
+        return OMX_ErrorUndefined;
+    }
+    cbd->post(msg);
 
     return OMX_ErrorNone;
 }
@@ -448,7 +458,12 @@ OMX_ERRORTYPE OMX::OnFillBufferDone(
     msg.u.extended_buffer_data.platform_private = pBuffer->pPlatformPrivate;
     msg.u.extended_buffer_data.data_ptr = pBuffer->pBuffer;
 
-    findDispatcher(node)->post(msg);
+    sp<OMX::CallbackDispatcher> cbd = findDispatcher(node);
+    if (cbd == NULL) {
+        LOGE("CallbackDispatcher for this node is not found");
+        return OMX_ErrorUndefined;
+    }
+    cbd->post(msg);
 
     return OMX_ErrorNone;
 }
-- 
1.7.4.1

